tools/wavetool.py

```diff
@@ -0,0 +1,157 @@
+    parser.add_argument('-p', '--preset', type=str, help='Commonly used wave presets\n'
+    'Supported presets are: sine_1k_16bit_48k_10s')
+    parser.add_argument('-c', '--compare', type=str, help='compare recorded wave and reference wave\n'
+    'parameter format: "mode;recorded_wave_path; reference_wave_path"\n'
+    'Supported comparison modes are: HWEchoRef')
+    return vars(parser.parse_args())
+
+def parse_params_and_gen_wav(cmd_param):
+    supported_type = ['sine', 'cosine']
+    param_keys = ['type', 'amp', 'freq', 'phase', 'sample_rate', 'channel', 'format', 'duration']
+    split_cmd_params = cmd_param.split(';')
+    wave_params = [param.strip() for param in split_cmd_params]
+    param_dict = dict(zip(param_keys, wave_params))
+    if param_dict['type'] not in supported_type:
+        print("Unsupported wave type")
+        sys.exit(1)
+    if param_dict['type'] in ['sine', 'cosine']:
+        wave_data = generateSinusoid(**param_dict)
+        return wave_data, param_dict
+
+def do_wave_compare(cmd_args):
+    supported_mode = ['HWEchoRef']
+    cmd_line = cmd_args['compare'].strip().split(';')
+    comparison_mode = cmd_line[0].strip()
+    if comparison_mode not in supported_mode:
+        print("Unsupported comparison mode: %s" % comparison_mode)
+        print('Supported comparison modes are: ' + str(supported_mode))
+        sys.exit(1)
+    wav_path, ref_path = [path.strip() for path in cmd_line[1:]]
+    if not os.path.exists(wav_path):
+        print('Recorded wave path does not exist: %s' % wav_path)
+        sys.exit(1)
+    if not os.path.exists(ref_path):
+        print('Reference wave path does not exist: %s' % ref_path)
+        sys.exit(1)
+    if comparison_mode == 'HWEchoRef':
+        compare_wav_hw_echo_ref(wav_path, ref_path)
+
+def save_wave(wave_data, path, **wave_params):
+    supported_format = ['S8', 'S16', 'S32', 'F32']
+    wave_path = path
+    if path == '.' or not path.endswith('wav'):
+        wave_name = wave_params['type'] + '_' + wave_params['channel'] + 'ch_' + wave_params['freq'] + 'Hz_' \
+        + wave_params['sample_rate'] + '_' + wave_params['format'] + '.wav'
+        wave_path = path + '/'  + wave_name
+    wave_format = wave_params['format'].upper()
+    if wave_format not in supported_format:
+        print('Unsupported wave format: %s' % wave_format)
+        sys.exit(1)
+    if wave_format == 'S8':
+        wave_data = (np.iinfo(np.int8).max * wave_data).astype(np.int8)
+    if wave_format == 'S16':
+        wave_data = (np.iinfo(np.int16).max * wave_data).astype(np.int16)
+    if wave_format == 'S32':
+        wave_data = (np.iinfo(np.int32).max * wave_data).astype(np.int32)
+    try:
+        wavefile.write(wave_path, int(wave_params['sample_rate']), wave_data)
+    except:
+        print("Path specified not valid: %s" % wave_path)
+
+# remove zeros in two sides
+def trim_wave(wave):
```

Comment:
It looks like `numpy` already has a `trim_zeroes()` function?

Reply:
1. trim_zeros() in numpy only applies to 1-D arrary. and our audio is 2-D array.
2. Once wave goes through DAC/ADC, the zero may become a small value close to zero, eg, 15, -48.  (This is the situation we met in WOV test)
I need to keep self-implemented function.

Reply:
1. `wave[:,0]` is a 1-D array, you can use `trim_zeros()` on it.
2. Where is this rounding? I don't this it in this function. 2. bis On line 115 you mentioned a bit-for-bit comparison. I'm confused and also a bit worried because rounding if any should be one of the most visible things  / main parameter in this code and I haven't spotted it yet. It's also a sure way to get different results with duplicated test code, see @plbossart's comment.

Reply:
Some rounding just got added, see below.